libs/compute_cuda/include/hpx/compute/cuda/target.hpp

```diff
@@ -32,6 +35,149 @@
 #include <hpx/config/warnings_prefix.hpp>
 
 namespace hpx { namespace compute { namespace cuda {
+
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail {
+        struct runtime_registration_wrapper
+        {
+            runtime_registration_wrapper(hpx::runtime* rt)
+              : rt_(rt)
+            {
+                HPX_ASSERT(rt);
+
+                // Register this thread with HPX, this should be done once for
+                // each external OS-thread intended to invoke HPX functionality.
+                // Calling this function more than once on the same thread will
+                // report an error.
+                hpx::error_code ec(hpx::lightweight);    // ignore errors
+                hpx::register_thread(rt_, "cuda", ec);
+            }
+            ~runtime_registration_wrapper()
+            {
+                // Unregister the thread from HPX, this should be done once in
+                // the end before the external thread exists.
+                hpx::unregister_thread(rt_);
+            }
+
+            hpx::runtime* rt_;
+        };
+
+        template <typename Allocator>
+        struct future_data;
+
+        template <typename Allocator>
+        struct release_on_exit
+        {
+            release_on_exit(future_data<Allocator>* data)
+              : data_(data)
+            {
+            }
+
+            ~release_on_exit()
+            {
+                // release the shared state
+                lcos::detail::intrusive_ptr_release(data_);
+            }
+
+            future_data<Allocator>* data_;
+        };
```

Comment:
If we moved the functions below to the implementation file, then the two classes above could go there as well.

Reply:
The first class can be moved unchanged (after the function `stream_callback` was moved), the second class requires some changes in order to be moved however. Instead of referring to `future_data<Allocator>` it could be referring to the base class `future_data_refcnt_base` (https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/lcos/detail/future_data.hpp#L69).